### PR TITLE
Adds Bayesian plots to Python

### LINKED ISSUE #45: Add additional plots
We have a plotting module that contains a plotting figure class and the `plot_ref_sld` routine. We want to add the following, adapted from the MATLAB code:

- [x] `cornerPlot`
- [x] `plotHists`
- [x] `bayesShadedPlot`
- [x] `histp`
- [x] `mcmcplot`
- [x] ~~`shade`~~
- [x] `plotBayes` (this function simply calls other plotting functions)

### PR DESCRIPTION
**Pull request by alexhroom**
This PR fixes #45 by adding the remaining plotting functionality from RAT to the Python API. These are:

- `cornerPlot` -> `plot_corner`
- `plotHists` -> `plot_hists`
- `bayesShadedPlot` -> new `bayes` parameter for `plot_ref_sld`
- `histp` -> `plot_one_hist` (with `estimated_density` parameter for estimated density overplot)
- `mcmcplot` -> `plot_chain`
- `plotBayes` -> `plot_bayes` (just calls `plot_ref_sld` with and without confidence intervals, `plot_hists` and `plot_corner`)

### CODE REVIEW

**Review by DrPaulSharp — COMMENTED** [review]
The plots are looking good so far, and I'm really pleased you've thought at each stage how to make them look as good as possible and add extra features.

There are two main jobs for you to do: add tests and address the issue with normalisation (corner is not playing ball unfortunately).

Have a look through the other comments too, hopefully there's some ideas there that will turn the plots up to 11.

File: RATapi/utils/plotting.py [hunk 1]
```diff
@@ -62,10 +76,9 @@ def plot_ref_sld_helper(data: PlotEventData, fig: Optional[plt.figure] = None, d
     elif len(fig.axes) != 2:
         fig.clf()
         fig.subplots(1, 2)
-    fig.subplots_adjust(wspace=0.3)
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:65]: ah i'd accidentally removed that when i merged badly and thought i'd put it back in!

File: RATapi/utils/plotting.py [hunk 2]
```diff
@@ -173,9 +203,37 @@ def plot_ref_sld(
...
 
+    if bayes:
+        try:
+            # the predictionIntervals data consists of 5 rows:
+            # row 0: min with 95% confidence
+            # row 1: min with 65% confidence
+            # row 2: mean
+            # row 3: max with 65% confidence
+            # row 4: max with 95% confidence
+            interval = [0, 4] if bayes == 95 else [1, 3]
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:214]: sure: was replicating the behaviour of `bayesShadedPlots` but of course that's a separate function to the unshaded plotting in that case

[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:214]: well the current behaviour is that it reports 65% for any possible input except `95` OR `None` (default) - this line only runs if Bayes is actually given a value. should we warn/error the user if they haven't put in either of those options since it's either undefined or a mistake?

[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RATapi/utils/plotting.py:214]: I was thinking about this a bit more last night and yes, that's the right way to go.

File: RATapi/utils/plotting.py [hunk 3]
```diff
@@ -222,3 +280,232 @@ def __exit__(self, _exc_type, _exc_val, _traceback):
...
+            bins[:-1],
+            bins,
+            weights=counts,
+            edgecolor="black",
+            linewidth=1.2,
+            color="white",
+        )
+        axes.set_title(fit_names[i])
+
+        if estimated_density:
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py]: this is the intended behaviour (matches MATLAB) - will discuss on Monday whether it's what we want the intended behaviour to be

**Review by DrPaulSharp — COMMENTED** [review]
The plots are looking in great shape and the tests look good. I have a few more specific comments, plus a few more general things to think about to make this module as strong as possible:

- I think we need to consider the naming convention of plotting routines. I think rather than e.g., "plot_hist" and "plot_hist_panel" we should have "plot_one_hist" and "plot_hists"  (of course "plot_hists" and similar can plot a single example with one parameter index provided). This better matches the matlab and gives the simpler name to the more commonly used routines.
- Please add a test for plot_ref_sld with shading
- When running test_plotting I get the warning
```
RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`
```
I'd suggest explicitly closing figures generated by the tests, hopefully they will run quicker too. Also consider patching routines that generate figures if the actual figures are not required.

I'll try and establish what's happening with the live plots and matplotlib errors that we discussed.

File: RATapi/utils/plotting.py [hunk 4]
```diff
@@ -222,3 +290,474 @@ def __exit__(self, _exc_type, _exc_val, _traceback):
...
+    if params is None:
+        params = range(0, len(results.fitNames))
+    else:
+        params = list(map(name_to_index, params))
+
+    # defaults are applied inside each function - just pass blank dicts for now
+    if hist_kwargs is None:
+        hist_kwargs = {}
+    if hist2d_kwargs is None:
+        hist2d_kwargs = {}
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:366]: [it's generally considered bad practice to use mutable objects as default arguments](https://web.archive.org/web/20200221224620id_/http://effbot.org/zone/default-values.htm) - i don't know whether the problem happens in this function specifically but it's safer to do it this way if we edit the function in future

File: RATapi/utils/plotting.py [hunk 5]
```diff
@@ -222,3 +290,474 @@ def __exit__(self, _exc_type, _exc_val, _traceback):
...
+    else:
+        params = list(map(name_to_index, params))
+
+    if estimated_density is not None:
+        temp = {}  # as dictionary cannot change size during iteration
+        for key, value in estimated_density.items():
+            temp[name_to_index(key)] = value
+        estimated_density = temp
+    else:
+        estimated_density = {}
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:673]: to do the same density on all plots one could do e.g. `{param: 'kernel' for param in problem.parameters}` - my suggestion would be to point out that construction in a tutorial but if you think that's a bit too complex for users i could add some code that looks for an 'all' key and applies it to all params?

File: RATapi/utils/plotting.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RATapi/utils/plotting.py:366]: Ahh yes. Pydantic solves this problem, so that's why these arguments are used elsewhere in the code, but let's leave these here.

File: RATapi/utils/plotting.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RATapi/utils/plotting.py:673]: I can imagine Arwel's response . . .
Dict comprehension is fine for you and me, but yeah, I do think its best to do something simpler. An option that applies for all by default, with a dict input as an option beyond I feel that is the way to go.

File: RATapi/utils/plotting.py [hunk 6]
```diff
@@ -222,3 +290,474 @@ def __exit__(self, _exc_type, _exc_val, _traceback):
...
+    def name_to_index(param: Union[str, int]):
+        """Convert parameter names to indices."""
+        if isinstance(param, str):
+            return results.fitNames.index(param)
+        return param
+
+    if params is None:
+        params = range(0, len(results.fitNames))
+    else:
+        params = list(map(name_to_index, params))
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:727]: i remember it breaking something, but it all seems to work now without the `list()`

File: RATapi/utils/plotting.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:673]: added the option to add a `'default'` key to the dict, which lets the user set a default estimated density for all parameters not otherwise specified

File: RATapi/utils/plotting.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:795]: nevermind - it breaks because the panel plot helper needs to get the length of the iterator for the number of plots - `range` has a length attribute but `map` does not!

**Review by DrPaulSharp — COMMENTED** [review]
I've a few specific comments, alongside those could you please review the test coverage - it's lacking in a couple of places, in particular estimated densities on histograms and bad input for indices on plots.

File: tests/test_plotting.py [hunk 7]
```diff
@@ -45,8 +50,22 @@ def domains_data() -> PlotEventData:
 def fig(request) -> plt.figure:
     """Creates the fixture for the tests."""
     plt.close("all")
-    figure = plt.subplots(1, 3)[0]
-    return plot_ref_sld_helper(fig=figure, data=domains_data() if request.param else data())
+    figure = plt.subplots(1, 2)[0]
+    return RATplot.plot_ref_sld_helper(fig=figure, data=domains_data() if request.param else data())
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply tests/test_plotting.py:54]: oh right, i assumed it was a typo. for that exact reason i think that it should be its own test rather than implicit so that we can tell that we're actually testing it!

File: RATapi/utils/plotting.py [hunk 8]
```diff
@@ -224,3 +303,530 @@ def __exit__(self, _exc_type, _exc_val, _traceback):
...
+
+    def plot_one_chain(axes: Axes, i: int):
+        axes.plot(range(0, nsimulations, skip), chain[:, i][0:nsimulations:skip])
+        axes.set_title(results.fitNames[i])
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:800]: (discussed: no due to interface bloat and the simplicity of the actual plotting function)

**Review by DrPaulSharp — APPROVED** [review]
Have a look at the final point on tidying up the estimated density code and then we're all done!

File: RATapi/utils/plotting.py [hunk 9]
```diff
@@ -224,3 +303,526 @@ def __exit__(self, _exc_type, _exc_val, _traceback):
...
+
+    if params is None:
+        params = range(0, len(results.fitNames))
+    else:
+        params = list(map(fitname_to_ind, params))
+
+    if estimated_density is not None:
+        if isinstance(estimated_density, str):
+            estimated_density = {"default": estimated_density}
+        default = estimated_density.get("default")
```
[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:713]: can't remove lines 716-717 because `pop` throws a `KeyError` if there is no key `default`

[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:713]: changed `get` to `pop` but you can't remove `estimated_density = {"default": estimated_density}` as if `estimated_density` is given as a string, you can't iterate over its items and the iterator breaks

[the remark this answers is not in the document]
Comment (alexhroom) [reply RATapi/utils/plotting.py:713]: can't do this because you iterate over `estimated_density.items()` either. although looking at it i'm becoming dissatisfied with how the loop looks generally. will try to come up with a neater way
